File: packages/contracts/src/health.ts
import { z } from "zod";

import { IdentifierSchema, IsoDateTimeSchema } from "./primitives";

export const SYSTEM_HEALTH_STATES = [
  "healthy",
  "degraded",
  "warning",
  "failed",
  "unknown",
] as const;

export const SystemHealthStateSchema = z.enum(SYSTEM_HEALTH_STATES);

export const SYSTEM_PLANES = ["control_plane", "execution_plane"] as const;

export const SystemPlaneSchema = z.enum(SYSTEM_PLANES);

export const SystemComponentHealthSchema = z
  .object({
    code: IdentifierSchema,
    labelTh: IdentifierSchema,
    plane: SystemPlaneSchema,
    state: SystemHealthStateSchema,
    detail: IdentifierSchema,
    observedAt: IsoDateTimeSchema,
  })
  .strict();

export const SystemHealthSnapshotSchema = z
  .object({
    capturedAt: IsoDateTimeSchema,
    components: z.array(SystemComponentHealthSchema).min(1),
  })
  .strict();

export type SystemHealthState = z.infer<typeof SystemHealthStateSchema>;
export type SystemComponentHealth = z.infer<typeof SystemComponentHealthSchema>;
export type SystemHealthSnapshot = z.infer<typeof SystemHealthSnapshotSchema>;

const stateSeverity: Record<SystemHealthState, number> = {
  healthy: 0,
  unknown: 1,
  degraded: 2,
  warning: 3,
  failed: 4,
};

export function deriveSystemHealthState(
  components: readonly SystemComponentHealth[],
): SystemHealthState {
  if (components.length === 0) return "unknown";
  return components.reduce<SystemHealthState>(
    (worst, component) =>
      stateSeverity[component.state] > stateSeverity[worst]
        ? component.state
        : worst,
    "healthy",
  );
}

File: apps/worker/src/aurum_worker/models/health.py
"""System health wire contracts and deterministic aggregation."""

from __future__ import annotations

from enum import StrEnum

from pydantic import AwareDatetime, model_validator

from aurum_worker.models.base import Identifier, WireModel


class SystemHealthState(StrEnum):
    HEALTHY = "healthy"
    DEGRADED = "degraded"
    WARNING = "warning"
    FAILED = "failed"
    UNKNOWN = "unknown"


class SystemPlane(StrEnum):
    CONTROL_PLANE = "control_plane"
    EXECUTION_PLANE = "execution_plane"


class SystemComponentHealth(WireModel):
    code: Identifier
    label_th: Identifier
    plane: SystemPlane
    state: SystemHealthState
    detail: Identifier
    observed_at: AwareDatetime


class SystemHealthSnapshot(WireModel):
    captured_at: AwareDatetime
    components: tuple[SystemComponentHealth, ...]

    @model_validator(mode="after")
    def require_components(self) -> SystemHealthSnapshot:
        if not self.components:
            raise ValueError("System health snapshot requires a component.")
        return self


_STATE_SEVERITY: dict[SystemHealthState, int] = {
    SystemHealthState.HEALTHY: 0,
    SystemHealthState.UNKNOWN: 1,
    SystemHealthState.DEGRADED: 2,
    SystemHealthState.WARNING: 3,
    SystemHealthState.FAILED: 4,
}


def derive_system_health_state(
    components: tuple[SystemComponentHealth, ...],
) -> SystemHealthState:
    """Return the worst reported state; an empty set is uncertain."""

    if not components:
        return SystemHealthState.UNKNOWN
    return max(components, key=lambda item: _STATE_SEVERITY[item.state]).state
